server/src/com/cloud/network/IpAddressManagerImpl.java

```diff
@@ -1698,6 +1698,22 @@ public String acquireGuestIpAddress(Network network, String requestedIp) {
 
     Random _rand = new Random(System.currentTimeMillis());
 
+    /**
+     * Get the list of public IPs that need to be applied for a static NAT enable/disable operation.
+     * Manipulating only these ips prevents concurrency issues when disabling static nat at the same time.
+     * @param staticNats
+     * @return The list of IPs that need to be applied for the static NAT to work.
+     */
+    public List<IPAddressVO> getStaticNatSourceIps(List<? extends StaticNat> staticNats) {
+        List<IPAddressVO> userIps = new ArrayList<>();
+
+        for (StaticNat snat : staticNats) {
+            userIps.add(_ipAddressDao.findById(snat.getSourceIpAddressId()));
+        }
```

Comment:
Executing queries in a ``for`` loop in unnecessarily expensive.  Could you please refactor this ``for`` loop into a new method on ``IPAddressDao`` that retrieves a list of IP addresses for a list of IPs?